testing/src/main/java/io/grpc/testing/GrpcResourceCleanupRule.java

```diff
@@ -0,0 +1,153 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.grpc.testing;
+
+import io.grpc.ManagedChannel;
+import io.grpc.Server;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.concurrent.TimeUnit;
+import javax.annotation.Nonnull;
+import javax.annotation.concurrent.NotThreadSafe;
+import org.junit.rules.ExternalResource;
+import org.junit.rules.TestRule;
+
+/**
+ * A JUnit {@link TestRule} that can register gRPC resources and manages its automatic release at
+ * the end of the test.
+ */
+@NotThreadSafe
+public final class GrpcResourceCleanupRule extends ExternalResource {
+
+  private final List<Resource> resources = new ArrayList<Resource>();
+
+  private RuntimeException exception;
+
+  /**
+   * Registers the given channel to the rule. Once registered, the channel will be automatically
+   * shutdown at the end of the test.
+   *
+   * <p>This method need be properly synchronized when used in multiple threads. This method must
+   * not be used during test tear down.
+   *
+   * @return the input channel
+   */
+  public <T extends ManagedChannel> T register(@Nonnull T channel) {
+    resources.add(new ManagedChannelResource(channel));
+    return channel;
+  }
+
+  /**
+   * Registers the given server to the rule. Once registered, the server will be automatically
+   * shutdown at the end of the test.
+   *
+   * <p>This method need be properly synchronized when used in multiple threads. This method must
+   * not be used during test tear down.
+   *
+   * @return the input server
+   */
+  public <T extends Server> T register(@Nonnull T server) {
+    resources.add(new ServerResource(server));
+    return server;
+  }
+
+  /**
+   * Releases all the registered resources.
+   */
+  @Override
+  protected void after() {
+    for (int i = resources.size() - 1; i >= 0; i--) {
+      resources.get(i).cleanUp();
+    }
+
+    for (int i = resources.size() - 1; i >= 0; i--) {
+      resources.get(i).awaitCleanupDone();
+    }
+
+    resources.clear();
+
+    if (exception != null) {
+      throw exception;
+    }
+  }
+
+  private interface Resource {
+    void cleanUp();
+
+    void awaitCleanupDone();
+  }
+
+  private final class ManagedChannelResource implements Resource {
+    final ManagedChannel channel;
+
+    ManagedChannelResource(ManagedChannel channel) {
+      this.channel = channel;
+    }
+
+    @Override
+    public void cleanUp() {
+      channel.shutdown();
+    }
+
+    @Override
+    public void awaitCleanupDone() {
+      if (exception != null) {
+        channel.shutdownNow();
+        return;
+      }
+
+      try {
+        channel.awaitTermination(1, TimeUnit.MINUTES);
```

Comment:
What's the benefit of `shutdownNow()` and `awaitTermination()`, say for 5s? Providing an orderly shutdown? But I think that is a non-goal. Assert `isTerminated`?

Reply:
The point is to clean-up more fully to reduce chances of one test impacting another. And possibly to notice more issues within a single test. I agree it is probably not necessary in this case, but that doesn't mean it may not be a good idea.

Are you suggesting we should just call `shutdown()` and not `shutdownNow()`? I could be okay with that if we failed the test if the test doesn't shut down promptly. Although I think we would still call `shutdownNow()` in that case, but only when we were already failing the test.

Reply:
After comparing the pros and cons of these options,  what I would like to do now is:
-  Automatic resource cleanup means cleaning up everything registered for the users including all the running/pending rpcs, regardless of existence of user-configured duration. That means we do in all cases `shutdownNow()` at the beginning of teardown. Don't want to do `shutdown()` in some cases but `shutdownNow()` in other cases.
-  Try the best to reduce chances of one test impacting another. Have a user-configurable total timeout (say 10s by default) awaiting all the resources released.